Compose Map - graphics overlay

shubham7109 opened the pull request:
Related issue: `#2948` 

### Description
PR to add the graphics overlay functionality to the compose map

### Testing
Here is a code snippet to test adding a graphic overlay on button click and clearing the graphic overlay. Modify the code to test other MutableList properties like removeAt, addAll.. etc.  
```kotlin
@Composable
fun MainScreen() {
    val arcGISMap by remember { mutableStateOf(ArcGISMap(BasemapStyle.ArcGISStreets)) }
    val graphicsOverlays = rememberGraphicsOverlayCollection()
    var offset by remember { mutableStateOf(0.0) }

    Map(
        modifier = Modifier.fillMaxSize(),
        arcGISMap = arcGISMap,
        graphicsOverlays = graphicsOverlays
    )

    Column {
        Button(onClick = {
            val lineGraphicsOverlay = createGraphicsOverlay(offset)
            graphicsOverlays.add(lineGraphicsOverlay)
            // increase Y offset for the next graphics overlay to be added
            offset += 10
        }) {
            Text(text = "Add graphics overlay")
        }

        Button(onClick = {
            graphicsOverlays.clear()
            offset = 0.0 // reset offset
        }) {
            Text(text = "Clear graphics overlay")
        }
    }
}

/**
 * Creates a GraphicOverlay across the equator and moves the Y [offset]
 */
fun createGraphicsOverlay(offset: Double): GraphicsOverlay {
    val polylineBuilder = PolylineBuilder(SpatialReference.wgs84()) {
        addPoint(Point(-60.0, 0.0 + offset))
        addPoint(Point(60.0, 0.0 + offset))
    }

    val lineSymbol = SimpleLineSymbol(SimpleLineSymbolStyle.Dash, Color.red, 4.0f)

    val graphicsOverlay = GraphicsOverlay().apply {
        graphics.add(Graphic(polylineBuilder.toGeometry(), lineSymbol))
    }

    return graphicsOverlay
}

```

puneet-pdx commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GraphicsOverlayMutableList.kt:
```diff
@@ -0,0 +1,91 @@
+
+/**
+ * A collection class to encapsulate the [GraphicsOverlay] list used by the [com.arcgismaps.toolkit.geocompose.MapView]
+ * @since 200.3.0
```
add empty line before @since

puneet-pdx commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GraphicsOverlayMutableList.kt:
```diff
@@ -0,0 +1,91 @@
+    /**
+     * Sealed class used to notify the compose MapView to update the GraphicOverlays on the
+     * type of [ChangedEvent].
+     * @since 200.3.0
```
add line before @since
https://github.com/androidx/androidx/blob/androidx-main/compose/docs/compose-component-api-guidelines.md#documentation-example

puneet-pdx commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GraphicsOverlayMutableList.kt:
```diff
@@ -0,0 +1,91 @@
+    internal val changed: SharedFlow<ChangedEvent> = _changed.asSharedFlow()
+
+    override fun add(index: Int, element: GraphicsOverlay) {
+        graphicsOverlays.add(element)
```
```suggestion
        graphicsOverlays.add(index, element)
```

puneet-pdx replied:
Also is this overload needed? We don't add the overlay in the mapView.graphicsOverlay at the given index. Shouldn't this list be reflecting what is in the mapView.graphicsOverlay?

puneet-pdx replied:
I think it should be associated with a `Inserted` `ChangedEvent` that takes the index and element.

puneet-pdx commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GraphicsOverlayMutableList.kt:
```diff
@@ -0,0 +1,91 @@
+        return graphicsOverlays[index]
+    }
+
+    override fun set(index: Int, element: GraphicsOverlay): GraphicsOverlay {
```
Is this setter needed ? underlying we are not setting, just adding.

puneet-pdx replied:
either we need to remove this or add a set `ChangedEvent` that takes the index and the element

puneet-pdx commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt:
```diff
@@ -127,6 +151,30 @@ public inline fun rememberLocationDisplay(
+    key: Any? = null,
+    crossinline init: GraphicsOverlayMutableList.() -> Unit = {}
+): GraphicsOverlayMutableList {
+    if (ArcGISEnvironment.applicationContext == null) {
```
This check is not required here. It was needed for `LocationDisplay` as we need the context to create it.

shubham7109 commented:
Thanks for the review @puneet-pdx, based on the comments the implementation didn't feel complete for various types of usages with MutableLists. 

I've added support for the following types of ChangedEvents:
- Added: `add(element: GraphicsOverlay)`
- Inserted: `add(index: Int, element: GraphicsOverlay)`
- Removed: `removeAt(index: Int)` and, `remove(element: GraphicsOverlay)`
- Updated: `set(index: Int, element: GraphicsOverlay)`
- Cleared: `clear()`

gunt0001 commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt:
```diff
@@ -127,6 +165,22 @@ public inline fun rememberLocationDisplay(
 
+/**
+ * Create and [remember] a [GraphicsOverlayMutableList].
+ * Checks that [ArcGISEnvironment.applicationContext] is set and if not, sets one.
```
This line needs to be removed.
```suggestion
```

gunt0001 commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt:
```diff
@@ -102,6 +100,46 @@ public fun MapView(
         }
     }
+
+    LaunchedEffect(graphicsOverlays) {
```
You could factor this whole LaunchedEffect out into a private composable function to keep the `MapView()` function readable. Something like:
```kotlin
@Composable
private fun GraphicsOverlayUpdater(
   graphicsOverlays: GraphicsOverlayMutableList
) {
   LaunchedEffect(graphicsOverlays) {
      ...
   }
}
```

gunt0001 commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GraphicsOverlayCollection.kt:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    /**
+     * Add a [graphicsOverlay] to the composable [MapView]'s graphics overlays.
```
```suggestion
     * Add a [graphicsOverlay] to this GraphicsOverlayCollection.
```

gunt0001 commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GraphicsOverlayCollection.kt:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    /**
+     * Remove a [graphicsOverlay] from the composable [MapView]'s graphics overlays.
```
```suggestion
     * Remove a [graphicsOverlay] from this GraphicsOverlayCollection.
```

gunt0001 commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GraphicsOverlayCollection.kt:
```diff
@@ -0,0 +1,106 @@
+    }
+
+    /**
+     * Returns the size of this instance.
```
```suggestion
     * Returns the number of graphics overlays in this GraphicsOverlayCollection.
```

gunt0001 commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GraphicsOverlayCollection.kt:
```diff
@@ -0,0 +1,106 @@
+        get() = graphicsOverlays.size
+
+    /**
+     * Clears the list of [GraphicsOverlay] from the [MapView]
```
```suggestion
     * Clears all graphics overlays from this GraphicsOverlayCollection.
```

gunt0001 commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GraphicsOverlayCollection.kt:
```diff
@@ -0,0 +1,106 @@
+     *
+     * @since 200.3.0
+     */
+    internal sealed class ChangedEvent(internal val element: GraphicsOverlay? = null) {
```
You could remove the property from the base class and move it to `Added` and `Removed` instead. That way you can declare it as non-nullable property on these individual sub classes.
```suggestion
    internal sealed class ChangedEvent() {
```

gunt0001 commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GraphicsOverlayCollection.kt:
```diff
@@ -0,0 +1,106 @@
+     * @since 200.3.0
+     */
+    internal sealed class ChangedEvent(internal val element: GraphicsOverlay? = null) {
+        class Added(element: GraphicsOverlay) : ChangedEvent(element)
```
```suggestion
        class Added(val element: GraphicsOverlay) : ChangedEvent()
```

gunt0001 commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GraphicsOverlayCollection.kt:
```diff
@@ -0,0 +1,106 @@
+     */
+    internal sealed class ChangedEvent(internal val element: GraphicsOverlay? = null) {
+        class Added(element: GraphicsOverlay) : ChangedEvent(element)
+        class Removed(element: GraphicsOverlay) : ChangedEvent(element)
```
```suggestion
        class Removed(val element: GraphicsOverlay) : ChangedEvent()
```

gunt0001 commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GraphicsOverlayCollection.kt:
```diff
@@ -0,0 +1,106 @@
+    internal sealed class ChangedEvent(internal val element: GraphicsOverlay? = null) {
+        class Added(element: GraphicsOverlay) : ChangedEvent(element)
+        class Removed(element: GraphicsOverlay) : ChangedEvent(element)
+        class Cleared : ChangedEvent()
```
```suggestion
        object Cleared : ChangedEvent()
```

gunt0001 commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GraphicsOverlayCollection.kt:
```diff
@@ -0,0 +1,106 @@
+ * @since 200.3.0
+ */
+public class GraphicsOverlayCollection :
+    Iterable<GraphicsOverlay> {
```
Could this move onto line 31?

gunt0001 commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt:
```diff
@@ -127,6 +165,23 @@ public inline fun rememberLocationDisplay(
+ * [init] will be called when the [GraphicsOverlayCollection] is first created to configure its
+ * initial state.
+ *
+ * @param key invalidates the remembered LocationDisplay if different from the previous composition
```
```suggestion
 * @param key invalidates the remembered GraphicsOverlayCollection if different from the previous composition
```

shubham7109 commented:
Thanks @gunt0001 for the helpful feedback. I've applied the changes, ready for your review.

gunt0001 commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/GraphicsOverlayCollection.kt:
```diff
@@ -0,0 +1,105 @@
+ *
+ * @since 200.3.0
+ */
+public class GraphicsOverlayCollection : Iterable<GraphicsOverlay> {
```
You can annotate this class with `@Stable`:
```suggestion
@Stable
public class GraphicsOverlayCollection : Iterable<GraphicsOverlay> {
```

gunt0001 commented on toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt:
```diff
@@ -109,6 +111,42 @@ public fun MapView(
+            when (changedEvent) {
+                // On GraphicsOverlay added:
+                is GraphicsOverlayCollection.ChangedEvent.Added ->
+                    changedEvent.element?.let { mapView.graphicsOverlays.add(it) }
```
I think the null check is redundant now:
```suggestion
                    mapView.graphicsOverlays.add(changedEvent.element)
```

gunt0001 reviewed (approved):
@shubham7109 - see final comments, otherwise looks good.

hud10837 reviewed (approved):
LGTM 👍
